probeinterface/probe.py

```diff
@@ -0,0 +1,259 @@
+        plane_axes:  (num_electrodes, 2, ndim)
+            This defines the electrod plane (2d or 3d)
+            
+        shapes: scalar or array in 'circle'/'square'/'rect'
+            Shape for each electrodes.
+        
+        shape_params dict or list of dict
+            Contain kargs for shapes ("radius" for circle, "width" for sqaure, "width/height" for rect)
+        """
+        assert positions is not None
+        
+        positions = np.array(positions)
+        if positions.shape[1] != self.ndim:
+            raise ValueErrorr('posistions.shape[1] and ndim do not match!')
+        
+        self.electrode_positions = positions
+        n = positions.shape[0]
+        
+        # This defines the electrod plane (2d or 3d) where the electrode lies.
+        # For 2D we make auto
+        if plane_axes is None:
+            if self.ndim ==3:
+                raise ValueError('you need to give plane_axes')
+            else:
+                plane_axes = np.zeros((n, 2, self.ndim))
+                plane_axes[:, 0, 0] = 1
+                plane_axes[:, 1, 1] = 1
+        plane_axes = np.array(plane_axes)
+        self.electrode_plane_axes = plane_axes
+
+        # shape
+        if isinstance(shapes, str):
+            shapes = [shapes] * n
+        shapes = np.array(shapes)
+        if not np.all(np.in1d(shapes, _possible_electrode_shapes)):
+            raise ValueError(f'Electrodes shape must be in {_possible_electrode_shapes}')
+        if shapes.shape[0] !=n:
+            raise ValueError(f'Electrodes shape must have same length as posistions')
+        self.electrode_shapes = np.array(shapes)
+        
+        # shape params
+        if isinstance(shape_params, dict):
+            shape_params = [shape_params] * n
+        self.electrode_shape_params = np.array(shape_params)
+    
+    def set_shape_vertices(self, shape_vertices):
+        shape_vertices = np.asarray(shape_vertices)
+        if shape_vertices.shape[1] != self.ndim:
+            raise ValueErrorr('shape_vertices.shape[1] and ndim do not match!')
+        self.probe_shape_vertices = shape_vertices
+    
+    def create_auto_shape(self, probe_type='rect', margin=20):
+        if self.ndim !=2:
+            raise NotImplementedError
+        
+        x0 = np.min(self.electrode_positions[:, 0])
+        x1 = np.max(self.electrode_positions[:, 0])
+        x0 -= margin
+        x1 += margin
+        
+        y0 = np.min(self.electrode_positions[:, 1])
+        y1 = np.max(self.electrode_positions[:, 1])
+        y0 -= margin
+        y1 += margin
+        
+        if probe_type == 'rect':
+            vertices = [(x0, y0), (x1, y0), (x1, y1), (x0, y1)]
+        elif probe_type == 'tip':
+            tip = ((x0+x1)*0.5, y0 - margin*4)
+            vertices = [(x0, y0), tip, (x1, y0), (x1, y1), (x0, y1)]
+        else:
+            raise ValueError()
+        self.set_shape_vertices(vertices)
+    
+    def set_device_channel_indices(self, channel_indices):
```

Comment:
Seems like we are using "electrode" and "channel" interchangeably. I would choose one of the two and be consistent I think. Probably "channel" since that is more inline with spikeinterface?

Reply:
I pefer to have theses 2 naming.
Electrode is more the physical stuff on the probe.
The channel is more related to the device channel.

This a main stuff that is not done correctly elsewhere: electrode have 2 index:

 *  the one from the probe manufacturor
  * the other one depending of the headstage amplififier and on which port you connected the headstage.

This is the case when neuronexus+openephys board the channel is a total mess.
But also for some other devices.